### PR TITLE
test(integration): add test for multiline logs in containerd log format

### PR DESCRIPTION
**Pull request by kkujawa-sumo**
##### Description

test(integration): add test for multiline logs in containerd log format

---

##### Checklist

Remove items which don't apply to your PR.

- [ ] Changelog updated

###### Testing performed

- run `make test TEST_NAME="Test_Helm_FluentBit_Containerd_Multiline_Logs"`

### CODE REVIEW

File: tests/integration/internal/stepfuncs/logs.go
Comment (pmalek-sumo) [inline tests/integration/internal/stepfuncs/logs.go:50]: 🤔

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [inline tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:46]: There's a PR that tries to address this: https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/2012 

Do you mind waiting with this until it gets merged :) ?

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [inline tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:90]: I'd remove all the code related to metrics both from the test code and from values.yaml.

This is already tested elsewhere. Just leave here whatever is related to containerd multiline logs.

I'd also remove the tests related to logs as that is tested also in a separate test.

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (pmalek-sumo) [inline tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml:144]: Do we need this for this test?

**Review by pmalek-sumo — CHANGES_REQUESTED** [review]

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (kkujawa-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:46]: I think that this PR can wait until #2012 is merged

File: tests/integration/internal/stepfuncs/logs.go
Comment (kkujawa-sumo) [reply tests/integration/internal/stepfuncs/logs.go:50]: Thanks for catching this, it is no longer needed.

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (kkujawa-sumo) [reply tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml]: OK, we can remove configuration for systemd logs, I removed also tests for non-multiline logs from `Test_Helm_FluentBit_Containerd_Multiline_Logs`

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (kkujawa-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:80]: done

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:36]: @kkujawa-sumo https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/2012 is merged 🎊

Comment (kkujawa-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go]: @pmalek-sumo please take another look

File: tests/integration/yamls/pod_multiline_long_lines.yaml
Comment (pmalek-sumo) [inline tests/integration/yamls/pod_multiline_long_lines.yaml:27]: Just for my memory: does that setup have this limitation that if we didn't send this very last log line then none of the previous logs would be flushed?

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [inline tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:29]: Maybe add some comments that these values are hardcoded in the pod definition yaml?

Or maybe even better: instead of applying a `.yaml` that's saved on disk we could define the pod in code and then we could have control over the loop counts and we wouldn't have to rely on "implicit dependency" between these `consts` here and the pod `.yaml` file?

WDYT?

Comment (swiatekm-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:29]: As a point of reference, we've defined a similar thing in code here: https://github.com/SumoLogic/sumologic-kubernetes-collection/blob/main/tests/integration/internal/logsgenerator/logsgenerator.go

File: tests/integration/yamls/pod_multiline_long_lines.yaml
Comment (kkujawa-sumo) [reply tests/integration/yamls/pod_multiline_long_lines.yaml:27]: In this solution buffering of log lines is controlled in `tail` input plugin so it s as good a previously used configurations, lua filter just cleans logs by removing information specific for containerd log format and joins multipart logs into single line (which are sent as part of multiline log).

I see that there is `Flush` parameter in `Service` section https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/configuration-file#config_section and it seems that it prevents the situation that nothing is sent.

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (pmalek-sumo) [reply tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml:13]: bump: @kkujawa-sumo Do you mind adding this?

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:29]: I've created an issue for this so that we can perhaps make this use less hardcoded and (very) loosely coupled `.yaml` files.

https://github.com/SumoLogic/sumologic-kubernetes-collection/issues/2055

Ideally we could use that approach elsewhere as well, e.g. in the place where you mentioned it @swiatekm-sumo

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (pmalek-sumo) [inline tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml:7]: As one last thing I'd ask to rebase this PR and remove this as well as the fluentd resource section since that is already covered by https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/2041

And then only override what's not set in common yaml.

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (kkujawa-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:29]: It will be good to have everything in the code but unfortunately size of arguments for processes is limited, related articles [1](https://www.in-ulm.de/~mascheck/various/argmax/), [2](https://www.baeldung.com/linux/argument-list-too-long-error) 

Following error occurred:
```
cat /var/log/containers/example-pod-multiline-logs-long-lines_multiline-logs-generator_example-pod-multiline-logs-long-lines-3cd7f6f3fd6f2a4eda236a9c55882beb05efe78f3e18e2695b09d7b635489433.log 
2022-02-22T13:57:55.662403778Z stderr F standard_init_linux.go:228: exec user process caused: argument list too long
```
and because of that we cannot pass explicitly very long strings  as arguments to `echo` and we cannot create very long number of arguments as `args` in Pod specification

Experiments with changing Pod in `yaml` to `go` are available on branch: https://github.com/SumoLogic/sumologic-kubernetes-collection/tree/kk-test-multiline-logs-trial-to-replace-yaml-with-go

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (kkujawa-sumo) [reply tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml:5]: Infromation is added [here](https://github.com/SumoLogic/sumologic-kubernetes-collection/pull/2015/files#diff-1e8f2418efe03ffd07e138c023c67c17153d806f70176c3f03426e088670240cR2)

File: tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml
Comment (kkujawa-sumo) [reply tests/integration/values/values_helm_fluentbit_containerd_multiline_logs.yaml]: done

File: tests/integration/helm_fluentbit_containerd_multiline_logs_test.go
Comment (pmalek-sumo) [reply tests/integration/helm_fluentbit_containerd_multiline_logs_test.go:29]: Thanks for checking if it's possible 🙏 

Ideally I'd like to have that documented somewhere, perhaps an issue? Perhaps a comment? So that we know we've investigated that and we know what are the limitations.

Maybe a "known limitations/issues" section in tests readme?

**Comment by kkujawa-sumo** [comment]
@pmalek-sumo please take another look

File: tests/integration/README.md
Comment (pmalek-sumo) [inline tests/integration/README.md:189]: 🚀

File: tests/integration/internal/constants.go
Comment (pmalek-sumo) [inline tests/integration/internal/constants.go:26]: Can we make it a bit more concrete/unique? So that in case we need to debug we don't have logs from other stuff mixed in the `default` namespace? We can call it e.g. `multiline-tests` or something like that

Comment (kkujawa-sumo) [reply tests/integration/internal/constants.go:26]: ok

**Pull request merged**
